### ISSUE TITLE
Hive Support

### ISSUE THREAD

**Issue by sengelbert** [mybatis/migrations#25]
Is Hive supported?

I have added all the jar files that I thought would be needed but I am still getting an error when trying to run the migrations. I am however able to list what migrations have not been run.

## List of hive jar driver files:
commons-codec-1.5.jar
commons-logging-1.1.1.jar
hadoop-common-2.6.0.jar
hive-common-0.13.0.jar
hive-jdbc-0.13.0.jar
hive-service-0.13.0.jar
httpclient-4.2.5.jar
httpcore-4.2.5.jar
libthrift-0.9.1.jar
log4j-1.2.17.jar
slf4j-api-1.7.5.jar
slf4j-log4j12-1.7.5.jar

## Sample migrate commands:
```
[dev@sandbox hive]$ migrate status
------------------------------------------------------------------------
-- MyBatis Migrations - status
------------------------------------------------------------------------
ID             Applied At          Description
================================================================================
log4j:WARN No appenders could be found for logger (org.apache.thrift.transport.TSaslTransport).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
20141202195919    ...pending...    test

------------------------------------------------------------------------
-- MyBatis Migrations SUCCESS
-- Total time: 0s
-- Finished at: Mon Aug 03 19:52:32 UTC 2015
-- Final Memory: 22M/480M
------------------------------------------------------------------------
[dev@sandbox hive]$ migrate up
------------------------------------------------------------------------
-- MyBatis Migrations - up
------------------------------------------------------------------------
log4j:WARN No appenders could be found for logger (org.apache.thrift.transport.TSaslTransport).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
========== Applying: 20141202195919_test.sql ===================================
--  Create Changelog
CREATE TABLE dev.mybatis_test(
  source string, 
  siteid string)

Error executing: CREATE TABLE dev.mybatis_test(
  source string, 
  siteid string)
.  Cause: java.sql.SQLException: Method not supported
------------------------------------------------------------------------
-- MyBatis Migrations FAILURE
-- Total time: 1s
-- Finished at: Mon Aug 03 19:52:36 UTC 2015
-- Final Memory: 20M/480M
------------------------------------------------------------------------

ERROR: Error executing command.  Cause: org.apache.ibatis.jdbc.RuntimeSqlException: Error executing: CREATE TABLE dev.mybatis_test(
  source string, 
  siteid string)
.  Cause: java.sql.SQLException: Method not supported
```

**Comment by sheikamp** [comment]
Not sure if you still need this, but I've just created a pull request (https://github.com/mybatis/migrations/pull/34) to add Hive compatibility to the Migrations framework.

**Comment by sengelbert** [comment]
@sheikamp That is great! I am definitely still interested in using mybatis with Hive. Do you have details of what jar files are needed for Hive and Mybatis from your tests?

**Comment by sheikamp** [comment]
I feared that the PR would be rejected, because it's breaking downward compatibility. I'll see if there's a way to fix that. Anyway, I've added you as a collaborator to my fork. Here's how I got it running:

1. Install migrations like described in the manual
2. Clone the fork
3. Run maven package
4. Replace the original jar (should be in /usr/local/mybatis/mybatis-migrations-${project.version}/lib) with the one just created (can be found in the "target" subfolder)
5. Add two new options to your environment properties, *changelog_insert* and *changelog_delete*. You can find an example here: https://github.com/sheikamp/migrations/blob/master/src/test/java/org/apache/ibatis/migration/example/environments/nosql.properties. This is necessary because Hive doesn't allow simple Inserts/Deletes.

As for the drivers, seems like you already figured that out :wink:

**Comment by sengelbert** [comment]
Thank you! I will give this a go and let you know if I have any issues (probably on the forked repo).

**Comment by hazendaz** [comment]
@sheikamp @sengelbert Where did this end up?  Does the fork work?  Is there any possibility of bringing that in without breaking the current implementation?  Should we just close out this issue possibly with just a link on the readme for hive support on the fork?

**Comment by harawata** [comment]
Hi guys!

Any action in Hive side?
Looking at the source code for Hive JDBC driver, the implementation seems to be very primitive.
Why don't you try building [my fork](https://github.com/harawata/hive) and see if it works?

```bash
$ git clone https://github.com/harawata/hive.git
$ cd hive/jdbc
$ mvn clean package
```

I only modified [one line](https://github.com/harawata/hive/commit/1910632442ad9b70b41b28e37596843d1b9f7d3f), but if it works and if you ask them nicely with a link to this ticket, the patch might be accepted ;D

**Comment by sheikamp** [comment]
Hi, sorry for the late reply. The fork is working fine, but I still don't see how to merge it without breaking anything. It really might be best to fix the Hive driver instead of working around that.

**Comment by harawata** [comment]
> It really might be best to fix the Hive driver instead of working around that.

That's exactly what I suggested ;-)
I just filed a new ticket on their JIRA: https://issues.apache.org/jira/browse/HIVE-13614
Hope it gets accepted! (voting might help)

**Comment by emacarron** [comment]
I have just voted on the issue.

**Comment by harawata** [comment]
Cool! Thanks Eduardo! =D

**Comment by h3adache** [comment]
lol their jira is throwing an error when voting.

**Comment by harawata** [comment]
I see your vote now, thanks, Tim!

**Comment by sheikamp** [comment]
:+1:

**Comment by harawata** [comment]
The patch was accepted for Hive 2.2.0 😊

**Comment by syvineckruyk** [comment]
sorry to resurrect an old thread ... with the driver issue corrected ... is it expected that migrations should work ? Or does it also require some of the changes from the closed/rejected pull request referenced above ?

**Comment by harawata** [comment]
@syvineckruyk 
The issue reported here was fixed in Hive ≥2.2.0 and I assumed that there were users who are using Migrations with Hive.
If you use Hive ≥2.2.0 and have some problem, please create a new issue with the details.
